File: java/src/test/java/io/mendirl/aventofcode/java2020/Day4Test.java
package io.mendirl.aventofcode.java2020;

import org.assertj.core.api.Assertions;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.net.URISyntaxException;
import java.nio.file.Files;
import java.nio.file.Path;

import static io.mendirl.aventofcode.java2020.Day4.calcul_step1;
import static io.mendirl.aventofcode.java2020.Day4.calcul_step2;
import static io.mendirl.aventofcode.java2020.Day4.valid_1;
import static io.mendirl.aventofcode.java2020.Day4.valid_2;
import static io.mendirl.aventofcode.java2020.Day4.validation;

class Day4Test {

    @Test
    void test_part_1() throws URISyntaxException, IOException {
        var uri = getClass().getResource("/day4.txt").toURI();
        var entries = Files.readAllLines(Path.of(uri));

        var result = calcul_step1(entries);

        System.out.println("the result is : " + result);

        Assertions.assertThat(result).isEqualTo(228);
    }

    @Test
    void test_part_2() throws URISyntaxException, IOException {
        var uri = getClass().getResource("/day4.txt").toURI();
        var entries = Files.readAllLines(Path.of(uri));

        var result = calcul_step2(entries);

        System.out.println("the result is : " + result);

        Assertions.assertThat(result).isEqualTo(175);
    }

    @Test
    void test_valid_1_1() {
        var input = "ecl:gry pid:860033327 eyr:2020 hcl:#fffffd byr:1937 iyr:2017 cid:147 hgt:183cm";

        var result = valid_1().test(input);

        Assertions.assertThat(result).isTrue();
    }

    @Test
    void test_valid_1_2() {
        var input = "iyr:2013 ecl:amb cid:350 eyr:2023 pid:028048884 hcl:#cfa07d byr:1929";

        var result = valid_1().test(input);

        Assertions.assertThat(result).isFalse();
    }

    @Test
    void test_valid_1_3() {
        var input = "hcl:#ae17e1 iyr:2013 eyr:2024 ecl:brn pid:760753108 byr:1931 hgt:179cm";

        var result = valid_1().test(input);

        Assertions.assertThat(result).isTrue();
    }

    @Test
    void test_valid_1_4() {
        var input = "hcl:#cfa07d eyr:2025 pid:166559648 iyr:2011 ecl:brn hgt:59in";

        var result = valid_1().test(input);

        Assertions.assertThat(result).isFalse();
    }

    @Test
    void test_validation_byr() {
        var result1 = VALIDATION.BYR.rule().test("2002");
        Assertions.assertThat(result1).isTrue();
        var result1_1 = validation("byr:2002");
        Assertions.assertThat(result1_1).contains(VALIDATION.BYR);

        var result2 = VALIDATION.BYR.rule().test("2003");
        Assertions.assertThat(result2).isFalse();
        var result2_1 = validation("byr:2003");
        Assertions.assertThat(result2_1).isEmpty();
    }

    @Test
    void test_validation_hgt() {
        var result1 = VALIDATION.HGT.rule().test("60in");
        Assertions.assertThat(result1).isTrue();
        var result1_1 = validation("hgt:60in");
        Assertions.assertThat(result1_1).contains(VALIDATION.HGT);

        var result2 = VALIDATION.HGT.rule().test("190cm");
        Assertions.assertThat(result2).isTrue();
        var result2_1 = validation("hgt:190cm");
        Assertions.assertThat(result2_1).contains(VALIDATION.HGT);

        var result3 = VALIDATION.HGT.rule().test("190in");
        Assertions.assertThat(result3).isFalse();
        var result3_1 = validation("hgt:190in");
        Assertions.assertThat(result3_1).isEmpty();

        var result4 = VALIDATION.HGT.rule().test("190");
        Assertions.assertThat(result4).isFalse();
        var result4_1 = validation("hgt:190");
        Assertions.assertThat(result4_1).isEmpty();
    }

    @Test
    void test_validation_hcl() {
            var result1 = VALIDATION.HCL.rule().test("#123abc");
        Assertions.assertThat(result1).isTrue();
        var result1_1 = validation("hcl:#123abc");
        Assertions.assertThat(result1_1).contains(VALIDATION.HCL);

        var result2 = VALIDATION.HCL.rule().test("#123abz");
        Assertions.assertThat(result2).isFalse();
        var result2_1 = validation("hcl:#123abz");
        Assertions.assertThat(result2_1).isEmpty();

        var result3 = VALIDATION.HCL.rule().test("123abc");
        Assertions.assertThat(result3).isFalse();
        var result3_1 = validation("hcl:123abc");
        Assertions.assertThat(result3_1).isEmpty();
    }


    @Test
    void test_validation_ecl() {
        var result1 = VALIDATION.ECL.rule().test("brn");
        Assertions.assertThat(result1).isTrue();
        var result1_1 = validation("ecl:brn");
        Assertions.assertThat(result1_1).contains(VALIDATION.ECL);

        var result2 = VALIDATION.ECL.rule().test("wat");
        Assertions.assertThat(result2).isFalse();
        var result2_1 = validation("ecl:wat");
        Assertions.assertThat(result2_1).isEmpty();
    }

    @Test
    void test_validation_pid() {
        var result1 = VALIDATION.PID.rule().test("000000001");
        Assertions.assertThat(result1).isTrue();
        var result1_1 = validation("pid:000000001");
        Assertions.assertThat(result1_1).contains(VALIDATION.PID);

        var result2 = VALIDATION.PID.rule().test("0123456789");
        Assertions.assertThat(result2).isFalse();
        var result2_1 = validation("pid:0123456789");
        Assertions.assertThat(result2_1).isEmpty();
    }


    @Test
    void test_valid_2_1() {
        var input = "pid:087499704 hgt:74in ecl:grn iyr:2012 eyr:2030 byr:1980 hcl:#623a2f";

        var result = valid_2().test(input);

        Assertions.assertThat(result).isTrue();
    }

    @Test
    void test_valid_2_2() {
        var input = "eyr:1972 cid:100 hcl:#18171d ecl:amb hgt:170 pid:186cm iyr:2018 byr:1926";

        var result = valid_2().test(input);

        Assertions.assertThat(result).isFalse();
    }

    @Test
    void test_valid_2_3() {
        var input = "hcl:#888785 hgt:164cm byr:2001 iyr:2015 cid:88 pid:545766238 ecl:hzl eyr:2022";

        var result = valid_2().test(input);

        Assertions.assertThat(result).isTrue();
    }

    @Test
    void test_valid_2_4() {
        var input = "iyr:2019 hcl:#602927 eyr:1967 hgt:170cm ecl:grn pid:012533040 byr:1946";

        var result = valid_2().test(input);

        Assertions.assertThat(result).isFalse();
    }
}

File: kotlin2021/src/test/kotlin/io/mendirl/adventofcode21/Day4Test.kt
package io.mendirl.adventofcode21

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class Day4Test : DayTest(3) {


    @Test
    fun `sample part 1`() {
        val input = listOf(
            "7,4,9,5,11,17,23,2,0,14,21,24,10,16,13,6,15,25,12,22,18,20,8,19,3,26,1",

            "22 13 17 11  0",
            " 8  2 23  4 24",
            "21  9 14 16  7",
            " 6 10  3 18  5",
            " 1 12 20 15 19",

            " 3 15  0  2 22",
            " 9 18 13 17  5",
            "19  8  7 25 23",
            "20 11 10 24  4",
            "14 21 16 12  6",

            "14 21 17 24  4",
            "10 16 15  9 19",
            "18  8 23 26 20",
            "22 11 13  6  5",
            " 2  0 12  3  7",
        )

        val result = Day4().step1(input)

        Assertions.assertEquals(198, result)
    }

    @Test
    fun `test part 1`() {
        val result = Day4().step1(inputs)

        Assertions.assertEquals(3959450, result)
    }


    @Test
    fun `sample part 2`() {
        val input = listOf(
            "7,4,9,5,11,17,23,2,0,14,21,24,10,16,13,6,15,25,12,22,18,20,8,19,3,26,1",
            "",
            "22 13 17 11  0",
            " 8  2 23  4 24",
            "21  9 14 16  7",
            " 6 10  3 18  5",
            " 1 12 20 15 19",
            "",
            " 3 15  0  2 22",
            " 9 18 13 17  5",
            "19  8  7 25 23",
            "20 11 10 24  4",
            "14 21 16 12  6",
            "",
            "14 21 17 24  4",
            "10 16 15  9 19",
            "18  8 23 26 20",
            "22 11 13  6  5",
            " 2  0 12  3  7",
        )

        val result = Day4().step2(input)

        Assertions.assertEquals(230, result)
    }

    @Test
    fun `test part 2`() {
        val result = Day4().step2(inputs)

        Assertions.assertEquals(7440311, result)
    }

}
